### PR TITLE
[FLOSS H2] Implement network retries

### PR DESCRIPTION
**Pull request by Eiskasten**
## Basics

- [x] Short descriptions of your changes are in the release notes
      (added as entry in `doc/news/_preparation_next_release.md` which
      contains `_(my name)_`)
      **Please always add something to the release notes.**
- [x] Details of what you changed are in commit messages
      (first line should have `module: short statement` syntax)
- [ ] References to issues, e.g. `close #X`, are in the commit messages.
- [ ] The buildservers are happy. If not, fix **in this order**:
  - [ ] add a line in `doc/news/_preparation_next_release.md`
  - [ ] reformat the code with `scripts/dev/reformat-all`
  - [ ] make all unit tests pass
  - [ ] fix all memleaks
- [ ] The PR is rebased with current master.

## Checklist

- [ ] I added unit tests for my code
- [ ] I fully described what my PR does in the documentation
      (not in the PR description)
- [ ] I fixed all affected documentation (see [Documentation Guidelines](https://www.libelektra.org/devgettingstarted/documentation))
- [ ] I fixed all affected decisions (see [Decision Process](https://www.libelektra.org/decisions/decision-process))
- [ ] I added code comments, logging, and assertions as appropriate (see [Coding Guidelines](https://www.libelektra.org/devgettingstarted/coding))
- [ ] I updated all meta data (e.g. README.md of plugins and [METADATA.ini](https://master.libelektra.org/doc/METADATA.ini))
- [ ] I mentioned [every code](/.reuse/dep5) not directly written by me in [reuse syntax](https://reuse.software/)

## Review

- [ ] Documentation is introductory, concise, good to read and describes everything what the PR does
- [ ] Examples are well chosen and understandable
- [ ] Code is conforming to [our Coding Guidelines](https://master.libelektra.org/doc/CODING.md)
- [ ] APIs are conforming to [our Design Guidelines](https://master.libelektra.org/doc/DESIGN.md)
- [ ] Code is consistent to [our Design Decisions](https://master.libelektra.org/doc/decisions)

## Labels

- [ ] Add the "work in progress" label if you do not want the PR to be reviewed yet.
- [ ] Add the "ready to merge" label **if the basics are fulfilled** and no further pushes are planned by you.

### CODE REVIEW

**Comment by Bujuhu** [comment]
## Review

### Documentation is introductory, concise, good to read and describes everything what the PR does
Commit messages are good, user facing documentation was added, but no issue is mentioned. Is this a new feature that was created without an underlying issue?

### Examples are well chosen and understandable
No new examples were provided

### Code is conforming to [our Coding Guidelines](https://master.libelektra.org/doc/CODING.md)
The newly introduced code is fine, but the function `gethostbyname` should be camelCase but is lowercase. As this is a preexisting issue, I wouldn't fix this as part of this PR
### APIs are conforming to [our Design Guidelines](https://master.libelektra.org/doc/DESIGN.md)
Not applicable, as `network` is a plugin

### Code is consistent to [our Design Decisions](https://master.libelektra.org/doc/decisions)
Adheres to logging and error formatting guidelines

**Review by joni1993 — APPROVED** [review]

**Comment by Eiskasten** [comment]
> Looks good now! heart
> 
> Only test cases still fail.
> 
> @Eiskasten did you have a case where it actually succeeded after retries without any delays in between?

Nothing to reproduce, but wouldn't it be the case if the request happened during any kind of handover such as routers or GSM cell?

**Comment by kodebach** [comment]
> wouldn't it be the case if the request happened during any kind of handover such as routers or GSM cell?

The man page says:

```
       TRY_AGAIN
              A temporary error occurred on an authoritative name
              server.  Try again later.
```

So it sounds like the error would be server-side, not a local connection change.

I think, there should be a delay between retries (ideally exponential backoff). The way we currently retry there will only be a few nanoseconds or maybe microseconds (only a handful of instructions in assembly) between consecutive `gethostbyname` calls. I doubt that would ever be enough time for the problem to resolve itself.

**Comment by Eiskasten** [comment]
You are right, I will implement a delay

**Comment by flo91** [comment]
jenkins build libelektra please

**Comment by flo91** [comment]
jenkins build libelektra please

**Comment by markus2330** [comment]
jenkins build libelektra please

File: src/plugins/network/network.c [hunk 1]
```diff
@@ -50,6 +50,9 @@ int elektraNetworkAddrInfo (Key * toCheck)
 	return 0;
 }
 
+#define NETWORK_ATTEMPTS 3
+#define NETWORK_SECONDS_BETWEEN_ATTEMPTS 5
```
Comment (markus2330) [inline src/plugins/network/network.c:54]: This is long delay (users might wait for it):

```suggestion
#define NETWORK_SECONDS_BETWEEN_ATTEMPTS 1
```

**Comment by flo91** [comment]
Great improvement, thank you!

**Pull request merged**